tcp/state.ml

```diff
@@ -123,42 +119,45 @@ module Make(Time:V1_LWT.TIME) = struct
       match s, i with
       | Closed, Passive_open -> Listen
       | Closed, Send_syn a -> Syn_sent a
+      | Closed, Recv_rst -> Closed
       | Listen, Send_synack a -> Syn_rcvd a
+      | Listen, Recv_rst -> Listen
       | Syn_rcvd _, Timeout -> t.on_close (); Closed
-      | Syn_rcvd _, Recv_rst -> Closed
-      | Syn_sent _, Timeout -> t.on_close (); Closed
-      | Syn_sent a, Recv_synack b-> if diffone b a then Established else Syn_sent a
       | Syn_rcvd a, Recv_ack b -> if diffone b a then Established else Syn_rcvd a
+      | Syn_rcvd _, Recv_rst -> Listen
+      | Syn_sent _, Timeout -> t.on_close (); Closed
+      | Syn_sent a, Recv_synack b -> if diffone b a then Established else Syn_sent a
+      | Syn_sent a, Recv_rstack b ->
+        if diffone b a then begin t.on_close (); Closed end
+        else Syn_sent a
       | Established, Recv_ack _ -> Established
       | Established, Send_fin a -> Fin_wait_1 a
       | Established, Recv_fin -> Close_wait
       | Established, Timeout ->  t.on_close (); Closed
-      | Established, Recv_rst -> t.on_close (); Reset
       | Fin_wait_1 a, Recv_ack b ->
-        if diffone b a then
-          let count = 0 in
-          Lwt.async (fun () -> finwait2timer t count fin_wait_2_time);
-          Fin_wait_2 count
-        else
+        if diffone b a then begin
+          Lwt.async (fun () -> finwait2timer t fin_wait_2_time);
+          Fin_wait_2
+            (* so the first time we enter fin_wait_2, we set the count to 0 and
+               spawn a waiting thread; the waiting thread only closes the
+               connection if the state still matches the number.  So if we keep
+               receiving ACKs, won't we never close? That seems incorrect. *)
```

Comment:
Out-of-date comment? Looks like you removed the counter.
@balrajsingh any idea what this counter scheme was for?